cli/src/commands/tag.ts

```diff
@@ -0,0 +1,44 @@
+import { Command, flags } from "@oclif/command";
+import { hashContractDefinition } from "../../../lib/src/tagging/hash";
+import { safeParse } from "../common/safe-parse";
+
+const ARG_API = "spot_contract";
+
+/**
+ * oclif command to generate a tag based on a Spot contract
+ */
+export default class Tag extends Command {
+  static description = "Generate a version tag based on a Spot contract";
+
+  static examples = ["$ spot mock api.ts"];
+
+  static args = [
+    {
+      name: ARG_API,
+      required: true,
+      description: "path to Spot contract",
+      hidden: false
+    }
+  ];
+
+  static flags = {
+    help: flags.help({ char: "h" }),
+    algorithm: flags.string({
+      description: "digest algorithm used to generate the hash"
+    })
+  };
+
+  async run() {
+    const {
+      args,
+      flags: { algorithm }
+    } = this.parse(Tag);
+    try {
+      const contract = safeParse.call(this, args[ARG_API]).definition;
+      const hash = hashContractDefinition(contract, { algorithm });
```

Comment:
Hello, I have a question

1. There are cases where the Spot contract will change, but the generated output like `openapi` won't  change, so it will generate a different hash but the output is not really the same

2. There are cases where the Spot implementation changes, but my Spot contract have no changes, but it will generate a new output, here I would like the hash to be different, but it won't (an example is from v0.2.8 to v0.2.9)

Should we do the checksum against generated output instead of the spot contract?

Reply:
hey @alephao. 

1. We're not aiming to do a checksum based on an openapi output here, but rather on an internal representation of the contract itself (in memory representation) as in contains data that would not be reflected in the openapi output - e.g. `tests` :

<img width="294" alt="Screen Shot 2019-06-14 at 10 03 18 am" src="https://user-images.githubusercontent.com/3967393/59475036-b97c2f80-8e8c-11e9-9589-1819feb7351a.png">

2. We acknowledege this but we don't expect the spot openapi codegen (templates) to change as it's pretty stable. The contract definition is our source of truth for now but we're keeping in mind that this will need to be improved. The point here is to ensure we generate ~accurate hashes for versionning our client sdks.